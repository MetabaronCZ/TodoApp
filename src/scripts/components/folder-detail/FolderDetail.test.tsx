import React from 'react';
import { t } from 'i18next';
import { Provider } from 'react-redux';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { FolderDetail } from 'components/folder-detail/FolderDetail';
import * as FolderFields from 'components/folder-detail/FolderFields';

import { client } from 'client';
import { Folder } from 'models/Folder';

import { mockStore } from 'test/store';
import { withMockedProviders } from 'test/component';

const testData: Folder = { id: '0', title: 'Test folder' };

const getFolderDetail = (data?: Folder, error?: boolean): JSX.Element => {
  const store = mockStore();
  return withMockedProviders(
    <Provider store={store}>
      <FolderDetail data={data} fetchError={error} />
    </Provider>,
  );
};

describe('components/folder-detail/FolderDetail', () => {
  it('should render folder create variant', () => {
    const fields = vi.spyOn(FolderFields, 'FolderFields');
    expect(fields).toBeCalledTimes(0);

    const tree = render(getFolderDetail());
    expect(tree.container).toMatchSnapshot();

    expect(tree.getByText(t('page.folderCreate'))).toBeInTheDocument();
    expect(tree.getByText(t('create'))).toBeInTheDocument();

    expect(fields).toBeCalledTimes(1);
    expect(fields).lastCalledWith(
      expect.objectContaining({
        loading: false,
        errors: {},
        fields: {
          title: '',
        },
      }),
      expect.anything(),
    );
  });

  it('should render folder detail variant', () => {
    const fields = vi.spyOn(FolderFields, 'FolderFields');
    expect(fields).toBeCalledTimes(0);

    const tree = render(getFolderDetail(testData));
    expect(tree.container).toMatchSnapshot();

    expect(tree.getByText(t('page.folderDetail'))).toBeInTheDocument();
    expect(tree.getByText(t('edit'))).toBeInTheDocument();

    expect(fields).toBeCalledTimes(1);
    expect(fields).lastCalledWith(
      expect.objectContaining({
        loading: false,
        errors: {},
        fields: {
          title: 'Test folder',
        },
      }),
      expect.anything(),
    );
  });

  it('should render fetch error', () => {
    const fields = vi.spyOn(FolderFields, 'FolderFields');
    expect(fields).toBeCalledTimes(0);

    const tree = render(getFolderDetail(testData, true));
    expect(tree.container).toMatchSnapshot();

    expect(tree.getByText(t('error.detailLoading'))).toBeInTheDocument();
    expect(tree.getByText(t('page.folderDetail'))).toBeInTheDocument();
    expect(tree.queryByRole('button')).not.toBeInTheDocument();

    expect(fields).toBeCalledTimes(0);
  });

  it('should create folder', () => {
    const mockedFolderCreate = vi.spyOn(client.folder, 'create');
    const fields = vi.spyOn(FolderFields, 'FolderFields');

    const tree = render(getFolderDetail());
    expect(mockedFolderCreate).toBeCalledTimes(0);

    const input = tree.container.querySelector('input');
    expect(input).toBeInTheDocument();

    const submitButton = tree.getByRole('button');
    expect(submitButton).toBeInTheDocument();

    // trigger form submit (validation error)
    fireEvent.click(submitButton);
    expect(mockedFolderCreate).toBeCalledTimes(0);

    if (fields.mock.lastCall) {
      expect(fields.mock.lastCall[0].errors?.title).toBeTruthy();
    }

    // type folder title
    if (input) {
      fireEvent.change(input, { target: { value: 'New folder' } });
    }

    // trigger form submit (validation OK)
    fireEvent.click(submitButton);
    expect(mockedFolderCreate).toBeCalledTimes(1);
    expect(mockedFolderCreate).toBeCalledWith({ title: 'New folder' });
  });

  it('should edit folder', () => {
    const mockedFolderEdit = vi.spyOn(client.folder, 'edit');

    const tree = render(getFolderDetail(testData));
    expect(mockedFolderEdit).toBeCalledTimes(0);

    const input = tree.container.querySelector('input');
    expect(input).toBeInTheDocument();
    expect(input?.value).toEqual('Test folder');

    const submitButton = tree.getByRole('button');
    expect(submitButton).toBeInTheDocument();

    // type folder title
    if (input) {
      fireEvent.change(input, { target: { value: 'Edited folder' } });
    }

    // trigger form submit
    fireEvent.click(submitButton);
    expect(mockedFolderEdit).toBeCalledTimes(1);
    expect(mockedFolderEdit).toBeCalledWith('0', { title: 'Edited folder' });
  });
});
