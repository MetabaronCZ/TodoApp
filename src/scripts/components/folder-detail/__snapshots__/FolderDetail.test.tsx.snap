// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/folder-detail/FolderDetail should render fetch error 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 18px;
  line-height: 30px;
  font-weight: normal;
  margin: 0;
}

.c5 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

.c1 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: calc(30px + 1px);
  border-bottom: 1px solid #c0c0c0;
  background: #f0f0f0;
}

.c6 {
  margin-top: 10px;
}

<div>
  <div>
    <h1
      class="c0"
    >
      Detail složky
    </h1>
    <div
      class="c1"
    >
      <div>
        <a
          class="c2"
          href="/folder/list"
        >
          <i
            class="c3"
          >
            ❮
          </i>
          <span
            class="c4"
          >
            Zpět
          </span>
        </a>
      </div>
    </div>
    <p
      class="c5 c6"
    >
      Data se nepodařilo načíst. Zkuste to prosím později.
    </p>
  </div>
</div>
`;

exports[`components/folder-detail/FolderDetail should render folder create variant 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c14 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c15 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
}

.c15:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c15:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c15:disabled {
  color: #c0c0c0;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 18px;
  line-height: 30px;
  font-weight: normal;
  margin: 0;
}

.c5 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  border-bottom: 1px solid #c0c0c0;
}

.c6:last-child {
  border-bottom: none;
}

.c9 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c12 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
}

.c12:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c12:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c12:disabled {
  color: #c0c0c0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: calc(30px + 1px);
  border-bottom: 1px solid #c0c0c0;
  background: #f0f0f0;
}

.c13 {
  border-top: 1px solid #c0c0c0;
}

<div>
  <div>
    <h1
      class="c0"
    >
      Nová složka
    </h1>
    <div
      class="c1"
    >
      <div>
        <a
          class="c2"
          href="/folder/list"
        >
          <i
            class="c3"
          >
            ❮
          </i>
          <span
            class="c4"
          >
            Zpět
          </span>
        </a>
      </div>
    </div>
    <ul
      class="c5"
    >
      <li
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <label
              class="c9"
              for="form-field-0"
            >
              Název složky:
            </label>
          </div>
          <div
            class="c10"
          >
            <div
              class="c11"
            >
              <input
                class="c12"
                id="form-field-0"
                maxlength="80"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      class="c13"
    >
      <button
        class="c14 c15"
        type="button"
      >
        <i
          class="c3"
        >
          +
        </i>
        <span
          class="c4"
        >
          Vytvořit
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`components/folder-detail/FolderDetail should render folder detail variant 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c14 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c15 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
}

.c15:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c15:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c15:disabled {
  color: #c0c0c0;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 18px;
  line-height: 30px;
  font-weight: normal;
  margin: 0;
}

.c5 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  border-bottom: 1px solid #c0c0c0;
}

.c6:last-child {
  border-bottom: none;
}

.c9 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c12 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
}

.c12:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c12:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c12:disabled {
  color: #c0c0c0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: calc(30px + 1px);
  border-bottom: 1px solid #c0c0c0;
  background: #f0f0f0;
}

.c13 {
  border-top: 1px solid #c0c0c0;
}

<div>
  <div>
    <h1
      class="c0"
    >
      Detail složky
    </h1>
    <div
      class="c1"
    >
      <div>
        <a
          class="c2"
          href="/folder/list"
        >
          <i
            class="c3"
          >
            ❮
          </i>
          <span
            class="c4"
          >
            Zpět
          </span>
        </a>
      </div>
    </div>
    <ul
      class="c5"
    >
      <li
        class="c6"
      >
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <label
              class="c9"
              for="form-field-1"
            >
              Název složky:
            </label>
          </div>
          <div
            class="c10"
          >
            <div
              class="c11"
            >
              <input
                class="c12"
                id="form-field-1"
                maxlength="80"
                type="text"
                value="Test folder"
              />
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      class="c13"
    >
      <button
        class="c14 c15"
        type="button"
      >
        <i
          class="c3"
        >
          ✎
        </i>
        <span
          class="c4"
        >
          Upravit
        </span>
      </button>
    </div>
  </div>
</div>
`;
